### PR TITLE
Proposal: `RouterProvider` for `Link` components

### PR DESCRIPTION
**Pull request by thomasdigby**
Open to comments/suggestions.

This new provider would allow the component library to render React Router or Next.js `Link` in place of `a`. We currently use `NavigateHandler` to mimic the SPA behaviour in atom-core, but there are a few weird edge-cases and bugs that can only really be resolved by using `Link` components directly. This context layer allows any `@atom-learning/components` component to access the "Link", and we can still combine that internally with logic for internal/external URLs and rendering `a` or `Link`.

~I named it `ComponentsProvider` as I wanted it to be fairly generic and we don't really have a name for our component library. Very happy to rename, tweak the structure, props, etc.~

Renamed to `RouterProvider` to be specific about its intent, and React Aria has a similarly named component for the same purpose https://react-spectrum.adobe.com/react-aria/routing.html

### Atom core
```jsx
import { Link } from 'react-router-dom'
import { RouterProvider } from '@atom-learning/components'

const Providers = ({ children }) => (
  <RouterProvider Link={({ href, ...props }) => <Link to={href} {...props} />}>
    {/* Other Providers */}
    {children}
  </RouterProvider>
)
```

### Marketing site

```jsx
import RouterLink from 'next/link'
import { RouterProvider } from '@atom-learning/components'

const Providers = ({ children }) => (
  <RouterProvider Link={RouterLink}>
    {/* Other Providers */}
    {children}
  </RouterProvider>
)
```

### CODE REVIEW

**Comment by thomasdigby** [comment]
@miaambt Tagging you if you have any q's about integration with Next.js router

**Comment by andoulla** [comment]
> Open to comments/suggestions.
> 
> This new provider would allow the component library to render React Router or Next.js `Link` in place of `a`. We currently use `NavigateHandler` to mimic the SPA behaviour in atom-core, but there are a few weird edge-cases and bugs that can only really be resolved by using `Link` components directly. This context layer allows any `@atom-learning/components` component to access the "Link", and we can still combine that internally with logic for internal/external URLs and rendering `a` or `Link`.
> 
> I named it `ComponentsProvider` as I wanted it to be fairly generic and we don't really have a name for our component library. Very happy to rename, tweak the structure, props, etc.
> 
> ```js
> import { Link } from 'react-router-dom'
> import { ComponentsProvider } from '@atom-learning/components'
> 
> const Providers = ({ children }) => (
>   <ComponentsProvider value={{ Link: ({ href, ...props }) => <Link to={href} {...props} /> }}>
>     {/* Other Providers */}
>     {children}
>   </ComponentsProvider>
> )
> ```

I have a few questions:
- Will we need to add this component at the app-base level, or next to each link, do we need to use the react-router link? What is the advice there?
- I'm wondering if this will affect any of the unit-tests in core as it might make mocking difficult (however we could do it in the new customRender that Marty created)

**Comment by thomasdigby** [comment]
> > Open to comments/suggestions.
> > This new provider would allow the component library to render React Router or Next.js `Link` in place of `a`. We currently use `NavigateHandler` to mimic the SPA behaviour in atom-core, but there are a few weird edge-cases and bugs that can only really be resolved by using `Link` components directly. This context layer allows any `@atom-learning/components` component to access the "Link", and we can still combine that internally with logic for internal/external URLs and rendering `a` or `Link`.
> > I named it `ComponentsProvider` as I wanted it to be fairly generic and we don't really have a name for our component library. Very happy to rename, tweak the structure, props, etc.
> > ```js
> > import { Link } from 'react-router-dom'
> > import { ComponentsProvider } from '@atom-learning/components'
> > 
> > const Providers = ({ children }) => (
> >   <ComponentsProvider value={{ Link: ({ href, ...props }) => <Link to={href} {...props} /> }}>
> >     {/* Other Providers */}
> >     {children}
> >   </ComponentsProvider>
> > )
> > ```
> 
> I have a few questions:
> 
>     * Will we need to add this component at the app-base level, or next to each link, do we need to use the react-router link? What is the advice there?
> 
>     * I'm wondering if this will affect any of the unit-tests in core as it might make mocking difficult (however we could do it in the new customRender that Marty created)

It will just be a single provider, which we can just replace in the existing slot that the `NavigateHandler` currently lives in. Everything else stays the same. You keep using `<Link />` and `<Button />` from `@atom-learning/components`.

The unit tests in core should be unaffected as we'll be rendering this in the test-providers.

File: lib/src/components/action-icon/ActionIcon.tsx [hunk 1]
```diff
@@ -154,13 +155,15 @@ export const ActionIcon: React.ForwardRefExoticComponent<ActionIconProps> =
...
+      const { Link: ExternalLink } = React.useContext(ComponentsContext)
+
       const INVALID_CHILDREN_MESSAGE = `A single ${Icon.displayName} component is permitted as a child of ${ActionIcon.displayName}`
 
       invariant(React.Children.count(children) === 1, INVALID_CHILDREN_MESSAGE)
 
       const optionalLinkProps = href
         ? ({
-            as: 'a',
+            as: isExternalUrl(href) ? 'a' : ExternalLink,
```
[the remark this answers is not in the document]
Comment (thomasdigby) [reply lib/src/components/action-icon/ActionIcon.tsx:166]: Oh yeah true, within this context, `Link` would be fine. I tried to rename it and keep it consistent to ensure there's no misunderstanding about our internal link component and the "provided" link component. Maybe I could rename it like that... 🤔

File: lib/src/components/button/Button.tsx [hunk 2]
```diff
@@ -187,24 +188,27 @@ type ButtonProps = Override<
...
-        ref={ref}
-      >
-        {isLoading ? (
-          <WithLoader size={rest.size}>{children}</WithLoader>
-        ) : (
-          children
-        )}
-      </StyledButton>
-    )
+    ({ children, as, href, isLoading = false, onClick, ...rest }, ref) => {
```
[the remark this answers is not in the document]
Comment (thomasdigby) [reply lib/src/components/button/Button.tsx:191]: Did I not add one for `Button`? I can add a test 👍

[the remark this answers is not in the document]
Comment (andoulla) [reply lib/src/components/button/Button.tsx:191]: this one and actionIcon don't have one, that's why I noticed

File: lib/src/components/action-icon/ActionIcon.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (andoulla) [reply lib/src/components/action-icon/ActionIcon.tsx:166]: what about `propLink` it clearly states that it is the link we pass in. or `contextualLink`

File: lib/src/components/button/Button.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (thomasdigby) [reply lib/src/components/button/Button.tsx:191]: `ActionIcon` does, but this doesn't.

**Review by Mhoog — APPROVED** [review]

**Pull request closed without merging**